### DISCUSSION TITLE
VGC format filter

### DISCUSSION THREAD

**Discussion by GregLyons** [GregLyons/poke-team#2] [Ideas]
@sdeiley suggested in #1 that VGC be added to the format filter. This is something I would've liked to have added, but I don't yet see a simple way to implement/one which would be sustainable for me to maintain. Currently I'm using the [Pokemon Showdown packages](https://www.smogon.com/forums/threads/pkmn-ps-unofficial-pok%C3%A9mon-showdown-packages.3671010/) (`data` and `dex` specifically) to perform tier validation on the Pokemon. Using it is basically as simple as passing in the Pokemon name and gen to get the tier (either Singles or Doubles). I unfortunately don't have time to keep up with the changes in the metagame or the changes in the VGC rules between years, so something like the PS package--where I at worst just need to update the package to use new tiers--is necessary for me to keep this app up-to-date with the ever-changing meta. It would also be possible to put all the necessary VGC data *in* the app, but someone else who's much more knowledgeable on VGC than I would need to implement AND maintain that.

That being said, I think a VGC feature would be pretty cool. We could probably put a "VGC" option in the Tier Filter, and then the dropdown would give the VGC years for the appropriate generation. This could contain VGC years and/or the current VGC letter rankings (A, B, etc.). I believe the relevant files for implementing this would be:

- [Here](https://github.com/GregLyons/poke-team/blob/main/src/utils/smogonLogic.ts) for converting `psID`s (the identifying string for working with the PS packages) to tiers;
- [Here](https://github.com/GregLyons/poke-team/blob/main/src/hooks/App/PokemonFilter.tsx) where the function `validatePokemon` is used wherever the Tier Filter applies (e.g. the Planner);
- [Here](https://github.com/GregLyons/poke-team/blob/main/src/hooks/App/TierFilter.tsx) for the reducer for the `tierFilter` object;
- [Here](https://github.com/GregLyons/poke-team/blob/main/src/components/ControlPanel/TierFilterForm/TierFilterForm.tsx) for modifying the TierFilter component.

So for anyone who would want to contribute to implementing this feature, I think those files would be the easiest to work in/add to.

**Comment by sdeiley** [comment]
In the past two years, the VGC ruleset changed 11 times, but the only relevant changes have been 1. Allowing "restricted" (basically box legendaries) or disallowing them 2. How many restricted can be on a team and 3. A top ten usage banlist. Finally, VGC has always disallowed "mythical" pokemon.

I think there's an 80/20 opportunity here where very little effort covers almost all VGC formats. I haven't looked at the DB model so I'm not sure the cost of this approach, but I would add a new Pokémon Metadata field called "restrictedVGC" and assign it to the 30 or so pokemon that are applicable. Notably, this status has never changed. TPCI has never added a not new Pokémon to this list or removed an old one, so this should be a one-time effort along with new game pokedex. Additionally, you'd do the same and have one for "mythical" as well.

For the VGC Tier, always filter mythicals and never filter restricteds. Then have a footnote or something on the restricted page when viewing it in the team builder that "VGC rules periodically change. Please check the latest Battle Spot Doubles ruleset to make sure this Pokémon is legal to use." I haven't had time to look at the validatePokemon​ logic, but I'm assuming another database call wouldn't be too tough.

The ban list only happened 1 month so I wouldn't implement something for that at this point.
________________________________
From: GregLyons ***@***.***>
Sent: Tuesday, April 12, 2022 6:27:35 AM
To: GregLyons/poke-team ***@***.***>
Cc: Shane Deiley ***@***.***>; Mention ***@***.***>
Subject: [GregLyons/poke-team] VGC format filter (Discussion #2)


@sdeiley<https://nam12.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fsdeiley&data=05%7C01%7C%7Cf928416be77340387fd208da1c883526%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C637853668585083550%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=tcYoIKyR0LNGMphu5f%2Be9hcuQvoFgmREU%2B%2FVD1%2FmgFQ%3D&reserved=0> suggested in #1<https://nam12.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FGregLyons%2Fpoke-team%2Fissues%2F1&data=05%7C01%7C%7Cf928416be77340387fd208da1c883526%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C637853668585083550%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=leStxdNOKyTNMU8r2NVi71RV4o%2F00YoJMNBwJJDA46k%3D&reserved=0> that VGC be added to the format filter. This is something I would've liked to have added, but I don't yet see a way to implement it which would be sustainable for me to maintain. Currently I'm using the Pokemon Showdown packages<https://nam12.safelinks.protection.outlook.com/?url=https%3A%2F%2Fwww.smogon.com%2Fforums%2Fthreads%2Fpkmn-ps-unofficial-pok%25C3%25A9mon-showdown-packages.3671010%2F&data=05%7C01%7C%7Cf928416be77340387fd208da1c883526%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C637853668585083550%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=n%2BKdSOIczfQise%2BckiDWedFpBcwx352hSihfB9jJYDc%3D&reserved=0> (data and dex specifically) to perform tier validation on the Pokemon. Using it is basically as simple as passing in the Pokemon name and gen to get the tier (either Singles or Doubles). I unfortunately don't have time to keep up with the changes in the metagame, let alone the changes in the VGC rules, so something like this--where I at worst just need to update the package to use new tiers--is necessary for me to keep this app up-to-date with the ever-changing meta. It would also be possible to put all the necessary VGC data in the app, but someone else who's much more knowledgeable on VGC than I would need to implement AND maintain that.

That being said, I think a VGC feature would be pretty cool. We could probably put a "VGC" option in the Tier Filter, and then the dropdown would give the VGC years for the appropriate generation. I believe the relevant files for implementing this would be:

  *   Here<https://nam12.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FGregLyons%2Fpoke-team%2Fblob%2Fmain%2Fsrc%2Futils%2FsmogonLogic.ts&data=05%7C01%7C%7Cf928416be77340387fd208da1c883526%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C637853668585083550%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=ZK%2Bbszs%2BD69uCj4LoNYFihAAguv46Lq27fP5giJ9HXY%3D&reserved=0> for converting psIDs (the identifying string for working with the PS packages);
  *   Here<https://nam12.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FGregLyons%2Fpoke-team%2Fblob%2Fmain%2Fsrc%2Fhooks%2FApp%2FPokemonFilter.tsx&data=05%7C01%7C%7Cf928416be77340387fd208da1c883526%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C637853668585083550%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=J3tvswibZdC0RBe4ratKRqMTu3tX%2FIsVVtdIMS3I5i0%3D&reserved=0> where the function validatePokemon is used wherever the Tier Filter applies (e.g. the Planner)
  *   Here<https://nam12.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FGregLyons%2Fpoke-team%2Fblob%2Fmain%2Fsrc%2Fcomponents%2FControlPanel%2FTierFilterForm%2FTierFilterForm.tsx&data=05%7C01%7C%7Cf928416be77340387fd208da1c883526%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C637853668585083550%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=1Cc2jnXgGm9fVaemP6HUBLXPVdIcPgdpRjkjJeWQeqw%3D&reserved=0> for modifying the TierFilter component.

**Reply by GregLyons** [reply]
OK, that makes sense. I think the 80/20 approach is the way to go. So I just need these two lists, right? 

- [Restricted](https://liquipedia.net/pokemon/Formats/VGC/2021/Series_10)
- [Mythical](https://bulbapedia.bulbagarden.net/wiki/Mythical_Pok%C3%A9mon)

Given that the restricted/mythical statuses haven't changed across generations yet, I actually think the simplest approach at this moment would actually be for me to just keep the two lists (sorted by name for fast lookup) in the app itself (maybe in a file in [`src/utils`](https://github.com/GregLyons/poke-team/tree/main/src/utils)) and check whether or not the Pokemon belongs to that list while running `validatePokemon` (assuming "VGC" is selected in the `TierFilter`).

I'll also probably add legendary/mythical/restricted flags to the database at some point in the future as well, and then add them as fields to the API in case anyone ever wants to use them when running PokeGQL themselves. The reason I prefer doing the former method is that, if I stored the flags in the database and then relied on fetching them, I'd need to modify a *lot* of GraphQL queries (basically any one that fetches data for Pokemon icons), and I'd need to modify the `PokemonIconDatum` data type (which is passed to `validatePokemon`), and so there would be some other places I'd need to modify as well. This wouldn't be *too* hard to do, but unless these flags change between generations, it wouldn't be necessary. It's also more consistent with how the other tiers are handled (in app itself, rather than stored in the database), and the lists are small enough (comparable in log(n) size to the list of tiers) that there shouldn't really be much additional performance cost.

I'll work on this in the coming days and should be able to get it done this week or early next.

**Comment by GregLyons** [comment]
How's this?

![vgc-planner-restricted-only](https://user-images.githubusercontent.com/89315619/163187544-6a41c7a8-cffd-40c3-82e6-7451e34a9a00.png)
![vgc-planner-all](https://user-images.githubusercontent.com/89315619/163187548-5bb97b0e-312c-4ea8-87a6-cf07cb57f97e.png)
![vgc-planner-other-only](https://user-images.githubusercontent.com/89315619/163187549-ae550987-0de8-4c12-8241-40e62244328b.png)

So if you want to look for only restricted Pokemon, you can de-select everything else. If you want only legal Pokemon, you can filter out mythicals, and if you want non-restricted, legal Pokemon, you can select only "Other". The tier filter applies to your cart as well, and these tiers show up in the Quick Search.

![vgc-behemoth](https://user-images.githubusercontent.com/89315619/163189042-a135a1c5-2516-4882-8461-2115e8058387.png)
![vgc-behemoth-none](https://user-images.githubusercontent.com/89315619/163189037-d2c8b128-7ecd-4e71-83cd-98e349ffafae.png)
![vgc-quick-search](https://user-images.githubusercontent.com/89315619/163187547-dedcf9b3-215c-47fa-8db2-7eae32709195.png)

List of mythicals and restricted Pokemon are in [this file](https://github.com/GregLyons/poke-team/blob/vgc-format/src/utils/vgcLogic.ts) in the `vgc-format` branch. I believe I got all the mega/primal/etc. forms as well, even though they aren't in the lists I posted last night. It looks like things are working well on my end, so if it all looks good then it's ready to go into the `main` branch.

**Comment by sdeiley** [comment]
Looks amazing!  Great work.  I hope I’m not coming across as critical BTW! Just providing some perspective/knowledge in case there are gaps!

From: GregLyons ***@***.***>
Reply-To: GregLyons/poke-team ***@***.***>
Date: Wednesday, April 13, 2022 at 6:19 AM
To: GregLyons/poke-team ***@***.***>
Cc: Shane Deiley ***@***.***>, Mention ***@***.***>
Subject: Re: [GregLyons/poke-team] VGC format filter (Discussion #2)


How's this?

[vgc-planner-restricted-only]<https://nam12.safelinks.protection.outlook.com/?url=https%3A%2F%2Fuser-images.githubusercontent.com%2F89315619%2F163187544-6a41c7a8-cffd-40c3-82e6-7451e34a9a00.png&data=05%7C01%7C%7C8f121a5dd6e04a24568708da1d504e83%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C637854527992769580%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=wrZuglDEzWmH%2FlEtuoDb3xAEbbaRMXKYYHYyC41VwwM%3D&reserved=0>
[vgc-planner-all]<https://nam12.safelinks.protection.outlook.com/?url=https%3A%2F%2Fuser-images.githubusercontent.com%2F89315619%2F163187548-5bb97b0e-312c-4ea8-87a6-cf07cb57f97e.png&data=05%7C01%7C%7C8f121a5dd6e04a24568708da1d504e83%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C637854527992769580%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=jMc62T7fm%2BjdCRNWM%2F2rl9Lmh7AtmrhEsDy8RYP2S5s%3D&reserved=0>
[vgc-planner-other-only]<https://nam12.safelinks.protection.outlook.com/?url=https%3A%2F%2Fuser-images.githubusercontent.com%2F89315619%2F163187549-ae550987-0de8-4c12-8241-40e62244328b.png&data=05%7C01%7C%7C8f121a5dd6e04a24568708da1d504e83%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C637854527992769580%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=LDWtLUQPKqRCvdG4nkvqfZR9mzc%2F9TM1nCD8VNvYS2k%3D&reserved=0>

So if you want to look for only restricted Pokemon, you can de-select everything else. If you want only legal Pokemon, you can filter out mythicals, and if you want non-restricted, legal Pokemon, you can select only "Other". The tier filter applies to your cart as well, and these tiers show up in the Quick Search.

[vgc-behemoth]<https://nam12.safelinks.protection.outlook.com/?url=https%3A%2F%2Fuser-images.githubusercontent.com%2F89315619%2F163189042-a135a1c5-2516-4882-8461-2115e8058387.png&data=05%7C01%7C%7C8f121a5dd6e04a24568708da1d504e83%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C637854527992769580%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=3qGAVCPzsJZgHrnffzobHgLXs5DTq5tt4%2BllxTqhasc%3D&reserved=0>
[vgc-behemoth-none]<https://nam12.safelinks.protection.outlook.com/?url=https%3A%2F%2Fuser-images.githubusercontent.com%2F89315619%2F163189037-d2c8b128-7ecd-4e71-83cd-98e349ffafae.png&data=05%7C01%7C%7C8f121a5dd6e04a24568708da1d504e83%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C637854527992769580%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=i0m2KZpBLd5HB8sDpAtVO0b5dWkRRBaTjxCuefaNRi0%3D&reserved=0>
[vgc-quick-search]<https://nam12.safelinks.protection.outlook.com/?url=https%3A%2F%2Fuser-images.githubusercontent.com%2F89315619%2F163187547-dedcf9b3-215c-47fa-8db2-7eae32709195.png&data=05%7C01%7C%7C8f121a5dd6e04a24568708da1d504e83%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C637854527992769580%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=9hh12p5EtKqAcbrTwyHgONq2kwMc5dqKokp3d92bUoI%3D&reserved=0>

**Reply by GregLyons** [reply]
On the contrary, your perspective is really helpful! As much as I know about the game mechanics themselves, I was *hoping* that people who are more in the competitive scene could identify potential small quality-of-life change like this. Explaining it in 80-20 terms also makes it a lot easier for me.

Anyway, I merged in the branch to `main`, so the change should now be live! Thanks again.
